### PR TITLE
feat: migrate to typescript & rollup

### PR DESCRIPTION
**Pull request by meabed**
I have kept the changes to the minimum :) 

- migrate to typescript 
- migrate build to rollup 
- eslint typescript support
- prettier
- FIX  handle number 0 in query params and added one more test 

Demo URL: https://codesandbox.io/s/priceless-smoke-qer8y
Test published LIB: https://www.npmjs.com/package/build-url-ts
UNPKG: https://unpkg.com/browse/build-url-ts@5.0.5/dist/

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 node_modules
 dist
+.idea
```
Comment (steverydz) [inline .gitignore:3]: IDE specific files should be added to your global gitignore, not in the project

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+};
```
Comment (steverydz) [inline jest.config.js:4]: Needs a newline at the end of the file

File: package.json [hunk 3]
```diff
@@ -2,13 +2,14 @@
...
-    "build": "webpack",
-    "lint-js": "eslint src/**/*.js",
-    "prepublishOnly": "webpack",
+    "build": "rollup -c ./rollup.config.js",
+    "lint-js": "eslint src/**/*.ts",
+    "prepublishOnly": "rollup -c ./rollup.config.js",
     "test": "jest",
-    "test-watch": "jest --watch"
+    "test-watch": "jest --watch",
+    "prettier": "npx prettier --write '**/*.{ts,tsx,css,scss}'"
```
Comment (steverydz) [inline package.json:12]: prettier is already a dependency of this project so `npx` shouldn't be needed, although I can't see where this is used?

File: package.json [hunk 4]
```diff
@@ -38,15 +39,20 @@
   },
   "homepage": "https://github.com/steverydz/build-url#readme",
   "devDependencies": {
-    "@babel/core": "7.11.4",
-    "@babel/preset-env": "7.11.0",
-    "babel-loader": "8.1.0",
+    "@rollup/plugin-typescript": "^5.0.2",
```
Comment (steverydz) [inline package.json:42]: please pin the dependency

File: package.json [hunk 5]
```diff
@@ -38,15 +39,20 @@
...
+    "@types/eslint": "7.2.2",
+    "@types/jasmine": "^3.5.14",
```
Comment (steverydz) [inline package.json:44]: please pin the dependency

File: package.json [hunk 6]
```diff
@@ -38,15 +39,20 @@
...
+    "@types/jest": "^26.0.10",
```
Comment (steverydz) [inline package.json:45]: please pin the dependency

File: package.json [hunk 7]
```diff
@@ -38,15 +39,20 @@
...
+    "@types/prettier": "2.1.0",
+    "@typescript-eslint/eslint-plugin": "3.10.1",
+    "@typescript-eslint/parser": "^3.10.1",
```
Comment (steverydz) [inline package.json:48]: please pin the dependency

File: package.json [hunk 8]
```diff
@@ -38,15 +39,20 @@
...
     "eslint": "7.7.0",
     "eslint-config-prettier": "6.11.0",
     "jest": "26.4.2",
     "prettier": "2.1.1",
-    "terser-webpack-plugin": "^4.1.0",
-    "webpack": "^4.44.1",
-    "webpack-cli": "^3.3.12"
+    "rollup": "^2.26.8",
```
Comment (steverydz) [inline package.json:53]: please pin the dependency

File: package.json [hunk 9]
```diff
@@ -38,15 +39,20 @@
...
+    "ts-jest": "^26.3.0",
```
Comment (steverydz) [inline package.json:54]: please pin the dependency

File: package.json [hunk 10]
```diff
@@ -38,15 +39,20 @@
...
+    "tslib": "^2.0.1",
```
Comment (steverydz) [inline package.json:55]: please pin the dependency

File: package.json [hunk 11]
```diff
@@ -38,15 +39,20 @@
...
+    "typescript": "^4.0.2"
```
Comment (steverydz) [inline package.json:56]: please pin the dependency

**Review by steverydz — CHANGES_REQUESTED** [review]
Thanks for taking the time to do this. I have some comments to address in the code, and the `npm run build` command gives an error (see TravisCI build details).

```bash
[!] Error: Unexpected token (Note that you need plugins to import files that are not JavaScript)
```

**Comment by meabed** [comment]
Thanks for the comments :) I have fixed them! and it looks good now.

**Comment by meabed** [comment]
Sounds good :) Thanks for the support!

**Pull request closed without merging**